tests/pubsub/testgossipsub.nim

```diff
@@ -928,3 +930,74 @@ suite "GossipSub":
 
     await allFuturesThrowing(nodesFut.concat())
 
+  proc initializeGossipTest(): Future[(seq[PubSub], GossipSub, GossipSub)] {.async.} =
+    let nodes = generateNodes(
+      2,
+      gossip = true,
+      overheadRateLimit = Opt.some((20, 1.millis)))
+
+    discard await allFinished(
+      nodes[0].switch.start(),
+      nodes[1].switch.start(),
+    )
+
+    await subscribeNodes(nodes)
+
+    proc handle(topic: string, data: seq[byte]) {.async, gcsafe.} = discard
+
+    let gossip0 = GossipSub(nodes[0])
+    let gossip1 = GossipSub(nodes[1])
+
+    gossip0.subscribe("foobar", handle)
+    gossip1.subscribe("foobar", handle)
+    await waitSubGraph(nodes, "foobar")
+
+    return (nodes, gossip0,  gossip1)
+
+  proc getCurrentDisconnections(): float64 =
+    try:
+      libp2p_gossipsub_peers_rate_limit_disconnections.valueByName("libp2p_gossipsub_peers_rate_limit_disconnections_total", @["nim-libp2p"])
+    except KeyError:
+      0
+
+  asyncTest "e2e - GossipSub should process valid messages":
+    let (nodes, gossip0, gossip1) = await initializeGossipTest()
+
+    gossip0.broadcast(gossip1.mesh["foobar"], RPCMsg(
+      messages: @[Message(topicIDs: @["foobar"], data: "Valid data".toBytes)])
+      )
+    await sleepAsync(300.millis)
+
+    check getCurrentDisconnections() == 0
+
+    await stopNodes(nodes)
+
+  asyncTest "e2e - GossipSub should rate limit undecodable messages above what is allowed":
+    let disconnections = getCurrentDisconnections()
+
+    let (nodes, gossip0, gossip1) = await initializeGossipTest()
+
+    # Simulate sending an undecodable message
+    await gossip0.peers[gossip1.switch.peerInfo.peerId].sendEncoded(newSeqWith[byte](30, 1.byte))
+    await sleepAsync(300.millis)
+
+    check getCurrentDisconnections() == disconnections + 1
```

Comment:
Once this actually disconnects, we should check here that we are no longer connected
Can you add a TODO?

Otherwise, LGTM

Reply:
Did this and other minor improvements.